lib/datadog/lambda/metrics.rb

```diff
@@ -0,0 +1,66 @@
+# frozen_string_literal: true
+
+#
+# Unless explicitly stated otherwise all files in this repository are licensed
+# under the Apache License Version 2.0.
+#
+# This product includes software developed at Datadog (https://www.datadoghq.com/).
+# Copyright 2023 Datadog, Inc.
+#
+
+require 'datadog/statsd'
+
+module Datadog
+  # Metrics module contains the singleton class Client to send custom
+  # metrics to Datadog.
+  module Metrics
+    URL = 'localhost'
+    PORT = '8125'
+    # Client is a singleton class that instantiates a Datadog::Statsd
+    # client to send metrics to the Datadog Extension if present.
+    class Client
+      private_class_method :new
+
+      def self.instance
+        @instance ||= new
+      end
+
+      def distribution(name, value, time: nil, **tags)
+        tag_list = get_tags(**tags)
+
+        if Datadog::Utils.extension_running?
+          Datadog::Utils.logger.debug 'sending metrics through extension'
+          begin
+            @statsd.distribution(name, value, tags: tag_list)
+          rescue StandardError => e
+            Datadog::Utils.logger.warning "error sending metric to the extension: #{e}"
+          end
+        else
+          Datadog::Utils.logger.debug 'metrics are going to be handled by the forwarder'
+          time ||= Time.now
+          time_ms = time.to_f.to_i
+
+          metric = { e: time_ms, m: name, t: tag_list, v: value }
+          puts metric.to_json
+        end
+      end
+
+      def end
+        @statsd&.close
+      end
+
+      def get_tags(**tags)
+        tag_list = ["dd_lambda_layer:datadog-ruby#{Datadog::Lambda.dd_lambda_layer_tag}"]
+        tags.each do |tag|
+          tag_list.push("#{tag[0]}:#{tag[1]}")
+        end
+
+        tag_list
+      end
+
+      private_class_method def initialize
+        @statsd = Datadog::Statsd.new(URL, PORT, single_thread: true) if Datadog::Utils.extension_running?
+      end
```

Comment:
Maybe? As suggested by `dd-trace-rb` team, to be able to easily mock a module during testing.
```suggestion
      private_class_method def initialize(statsd = Datadog::Statsd.new)
        @statsd = statsd(URL, PORT, single_thread: true) if Datadog::Utils.extension_running?
      end
```

Reply:
rspec at least would allow you to do that mocking no problem even without the dependency injection.

Reply:
Yeah, I'm currently doing that. But that requires knowing that you have to mock all the time, making it harder to setup.

Reply:
I mean, with the dependency injection, you still gotta know that you have to mock it.

Reply:
@duncanista, I'd say give it a shot and see which one you like better.

Whatever looks easier to understand it likely the one you want to go with.

It's a trade-off between both options, so there's no pragmatic right or wrong answer here.